feat: Google Drive Picker

barredterra opened the pull request:
User can select files from Google Drive, using [Google Picker](https://developers.google.com/picker). The Google Drive icon is only shown if the Google integration is enabled. The behavior in the background is like attaching a URL. 

<img width="1421" alt="file_uploader" src="https://user-images.githubusercontent.com/14891507/115119122-1f650100-9fa7-11eb-89f0-d089da39c1d5.png">

If it's the first time, the user needs to sign into google and / or authorize frappe to access it. The access token is saved in `frappe.boot.user.google_drive_token` so this can be skipped most of the time. Also google remembers the authorization and will skip the authorization page in the future.

<img width="1417" alt="2021-03-30 Screenshot Sign in - Google Accounts" src="https://user-images.githubusercontent.com/14891507/112914611-adfb0680-90fc-11eb-876b-feed3e502292.png">

Frappe's FileUploader is closed and instead the Google Picker appears. The user can click on a file and confirm with "Select":

<img width="1417" alt="select_a_file" src="https://user-images.githubusercontent.com/14891507/112913929-13e68e80-90fb-11eb-87a3-a29852f9844b.png">

The selected file is added as an attachment to the doc:

<img width="1415" alt="file_attached" src="https://user-images.githubusercontent.com/14891507/115119143-30ae0d80-9fa7-11eb-8b3e-5347101f086c.png">

Google Settings:

![google_settings](https://user-images.githubusercontent.com/14891507/112989115-dae30400-9164-11eb-8a3f-91d1f3c31d2d.png)

Todo:

- [x] fix file preview / failed thumbnail requests
- [x] find a way to add nice file names
- [x] use the [new google drive logo](https://developers.google.com/drive/api/v3/branding)
- [x] set picker locale according to user language
- [x] separate section in **Google Settings**?
- [x] description for fields in **Google Settings**
- [x] tests
- [x] docs
- [x] remove questionable permissions introduced with https://github.com/frappe/frappe/pull/7625

If you find this useful, you're most welcome to contribute code or money to this feature.

Picker docs: https://developers.google.com/picker/docs
Docs: https://github.com/frappe/frappe_docs/pull/109

barredterra committed:
feat: google drive picker

barredterra committed:
fix: scope for google picker

barredterra committed:
fix: remove log statements

barredterra committed:
feat: better home view for google drive picker

barredterra committed:
feat: localize google drive picker

barredterra committed:
refactor: bind this

barredterra committed:
refactor: rename GoogleDrive -> GoogleDrivePicker

barredterra committed:
refactor: save one API call

surajshetty3416 commented:
@barredterra you can update your branch to avoid "indentation error" by sider.

barredterra committed:
Merge branch 'develop' into google_drive_picker

barredterra committed:
fix: spaces to tabs

barredterra committed:
fix: define global vars for sider

barredterra committed:
feat: separate switch for picker, descriptions

barredterra committed:
Merge branch 'develop' into google_drive_picker

barredterra committed:
test: google settings

barredterra committed:
test: picker enabled

barredterra committed:
feat: use nice file name

barredterra committed:
fix: new google drive logo

barredterra committed:
Merge branch 'develop' into google_drive_picker

barredterra commented:
Some of the failing tests are fixed with #12887

barredterra committed:
Merge branch 'develop' into google_drive_picker

barredterra commented:
Not stale

barredterra commented:
Not stale, waiting for review

barredterra commented:
Not stale

barredterra committed:
Merge branch 'develop' into google_drive_picker

barredterra commented:
@leela I'd appreciate your review on this PR :)

Alchez commented on frappe/handler.py:
```diff
@@ -144,8 +144,8 @@ def upload_file():
 	file_url = frappe.form_dict.file_url
 	folder = frappe.form_dict.folder or 'Home'
 	method = frappe.form_dict.method
+	filename = frappe.form_dict.get('file_name')
```
This is really a nitpick, but should this be consistent with the rest of the variables?

```suggestion
	filename = frappe.form_dict.file_name
```

Alchez commented on frappe/handler.py:
```diff
@@ -155,7 +155,7 @@ def upload_file():
 	frappe.local.uploaded_filename = filename
 
-	if frappe.session.user == 'Guest' or (user and not user.has_desk_access()):
+	if (frappe.session.user == 'Guest' or (user and not user.has_desk_access())) and not file_url:
```
If you make the `not file_url` check first, then the remaining expressions in the `and` clause won't be evaluated, which might save some calls.

Alchez commented on frappe/integrations/doctype/google_settings/google_settings.json:
```diff
@@ -18,10 +22,12 @@
    "label": "Enable"
   },
   {
+   "description": "The Client ID obtained from the Google Cloud Console under \"APIs & Services\" > \"Credentials\".",
```
Can we do links/hyperlinks in the description? Might be useful to just link to https://console.cloud.google.com/apis/credentials for quick access.

Alchez commented on frappe/integrations/doctype/google_settings/google_settings.py:
```diff
@@ -2,11 +2,29 @@
+def get_file_picker_settings():
+	"""Return all the data FileUploader needs to start the Google Drive Picker."""
+	if frappe.session.user == 'Guest':
+		return {'enabled': False}
```
Another nitpick, but I think Frappe's using Black now and is enforcing double-quotes for strings.

Alchez commented on frappe/integrations/doctype/google_settings/google_settings.py:
```diff
@@ -2,11 +2,29 @@
+	return "https://www.googleapis.com/oauth2/v4/token"
+
+
+@frappe.whitelist(allow_guest=True)
```
If the `allow_guest` is set to `False`, will that cause issues when trying to make requests on the frontend? Since this is private data being returned, I think we should be extra safe.

barredterra committed:
Merge branch 'develop' into google_drive_picker

barredterra committed:
refactor: suggestions from review

barredterra committed:
feat: links to google cloud console in description

barredterra commented:
Thanks a lot for your review, @Alchez. I made the suggested improvements:

- Fix style
- Disable guest access
- Add links to the appropriate places in Google Cloud Console to the field descriptions in **Google Settings**

barredterra committed:
test: change test cases after interface change

barredterra committed:
test: remove test_picker_as_guest

Guest access was disabled. Testing this would be equivalent to testing frappe.whitelist.

barredterra committed:
style: use double quotes

barredterra committed:
feat: remove "All" access from Google Settings

barredterra commented:
@surajshetty3416 , @Alchez I'd appreciate your review, if you have any time.

Alchez reviewed (approved):
Looks good to me!

revant commented:
https://board.castlecraft.co.in/improvement?name=GH-IMP-2021-00192

Komal-Saraf0609 commented:
Hi @barredterra,

Tried testing the above mentioned PR but after I click on the google drive option, it redirects me to the same page rather than taking me to the google drive of the logged in user.

Please refer the attached recording for the same:

https://user-images.githubusercontent.com/81952590/124557458-29382b00-de57-11eb-9d90-dc8216f41801.mov

Komal-Saraf0609 commented:
https://user-images.githubusercontent.com/81952590/124557458-29382b00-de57-11eb-9d90-dc8216f41801.mov

barredterra commented:
@Komal-Saraf0609 thanks for reviewing. Are you sure that your **Google Settings** are correct ([see docs](https://github.com/frappe/frappe_docs/pull/109))?

revant commented:
I'm getting this error?

![access-not-granted](https://user-images.githubusercontent.com/1040161/124792442-a3a59f80-df6a-11eb-8168-9a1cbd2f515b.png)

barredterra commented:
@revant, @Komal-Saraf0609 as long as this is not a Google-approved public app, every company will require their own credentials that are only valid internally. This is a restriction by google that we will have to live with.

Therefore we cannot all test on the same site / with the same credentials. Everyone will need to create and set their own.

revant commented:
I tried my credentials and everything works as expected.

Added a new note and attached a file

https://gh-imp-2021-00192.castlecraft.co.in/app/note/cool%20new%20note

Remember to add on google end

- Javascript uri
- redirect uri
- enable required scopes
- enable Google Drive API
- enable Google Picker API

Komal-Saraf0609 commented:
Hi,

Everything is working as expected for me as well.

It wasnt working because I didnt enable Google Picker API, after enabling it and adding it, everything started working as expected.

barredterra commented:
@mergifyio backport version-13-hotfix

sabgaby commented:
This works following instructions. However, atm it does not support Shared Drives; which we use exclusively. With [APIv3](https://developers.google.com/drive/api/v3/enable-shareddrives) `supportsAllDrives=true`.

barredterra commented:
@sabgaby looks like it's possible to enable shared drives for the Google Drive Picker. Would you like to sponsor this feature?

sabgaby commented:
@barredterra: sorry for the late reply 🙈 . We could consider sponsoring this feature. However, I don't know how that process works. Who does the coding? and how much does it cost? Guidance is appreciated!

barredterra commented:
@sabgaby feel free to reach out to ALYF (my company), Frappe Technologies or any other service provider / freelancer for a quote.
